exercises/linked-list/tests/linked-list.rs

```diff
@@ -0,0 +1,242 @@
+            self.0.set(num-1);
+        }
+    }
+
+    const N: usize  = 15;
+
+    let counter = Cell::new(N);
+    let mut list = LinkedList::new();
+    for _ in 0..N {
+        list.push_back(Counter(&counter));
+    }
+    assert_eq!(list.len(), N);
+    drop(list);
+    assert_eq!(counter.get(), 0);
+}
+
+
+#[test]
+fn clone_is_equal() {
+    let mut list = LinkedList::new();
+    for i in 0..10 {
+        list.push_back(i);
+    }
+    let mut list2 = clone_list(&list);
+
+    for _ in 0..10 {
+        assert_eq!(list.pop_back(), list2.pop_back());
+    }
+}
+
+#[test]
+fn insert_middle() {
+    let mut list = LinkedList::new();
+    for i in 0..10 {
+        list.push_back(i);
+    }
+    let mut list2 = clone_list(&list);
+
+    {
+        let mut cursor = list.cursor_tail();
+        for _ in 0..4 {
+            cursor.next();
+        }
+        //cursor.seek_forward(4);
+        while let Some(elem) = list2.pop_back() {
+            cursor.insert_after(elem);
+        }
+        //cursor.insert_list_after(&mut list2);
+    }
+
+    assert_eq!(list.len(), 20);
+    assert_eq!(list2.len(), 0);
+
+    let expected = (0..5).chain(0..10).chain(5..10);
+
+    for (exp, &actual) in expected.zip(list.iter()) {
+        assert_eq!(exp, actual);
+    }
+}
+
+#[test]
+fn head_tail_changes_on_push_back() {
+    let mut heads = vec![];
+    let mut tails = vec![];
+    let mut list = LinkedList::new();
+
+    for i in 0..10 {
+        list.push_back(i);
+        heads.push(list.cursor_head().peek_mut().map_or(std::ptr::null(), |r| r as *const i32));
+        tails.push(list.cursor_tail().peek_mut().map_or(std::ptr::null(), |r| r as *const i32));
+    }
+    tails.sort();
+    tails.dedup();
+
+    assert_eq!(tails.len(), 1);
+    assert_eq!(tails[0], heads[0]);
+
+    heads.sort();
+    heads.dedup();
+
+    assert_eq!(heads.len(), 10);
+}
+
+#[test]
+fn head_tail_changes_on_push_front() {
+    let mut heads = vec![];
+    let mut tails = vec![];
+    let mut list = LinkedList::new();
+
+    for i in 0..10 {
+        list.push_front(i);
+        heads.push(list.cursor_head().peek_mut().map_or(std::ptr::null(), |r| r as *const i32));
+        tails.push(list.cursor_tail().peek_mut().map_or(std::ptr::null(), |r| r as *const i32));
+    }
+    heads.sort();
+    heads.dedup();
+
+    assert_eq!(heads.len(), 1);
+    assert_eq!(heads[0], tails[0]);
+
+    tails.sort();
+    tails.dedup();
+
+    assert_eq!(tails.len(), 10);
+}
+
+#[test]
+fn linked_list_is_send_sync() {
+    trait AssertSend: Send {}
+    trait AssertSync: Sync {}
+
+    impl<T: Send> AssertSend for LinkedList<T> {}
+    impl<T: Sync> AssertSync for LinkedList<T> {}
+}
+
+#[allow(dead_code)]
+#[test]
+fn is_covariant() {
+    fn a<'a>(x: LinkedList<&'static str>) -> LinkedList<&'a str> {
+        x
+    }
+}
+
+#[test]
+fn is_generic() {
+    struct Foo;
+    LinkedList::<Foo>::new();
+}
+
+fn clone_list<T: Clone>(list: &LinkedList<T>) -> LinkedList<T> {
```

Comment:
Instead of implementing this function, what do you think of requiring students to implement [`FromIterator`](https://doc.rust-lang.org/std/iter/trait.FromIterator.html)? Then calls to `clone_list` could be replaced with `dest: LinkedList<_> = source.iter().cloned().collect();`.

Reply:
It would some more code to an already large exercise, but I'm not opposed. It's not hard to implement and it would make building arbitrary lists in the tests very clean.

Reply:
I like it because it's a useful trait that I don't think we require of students elsewhere.